docs(treesitter): docs and typing for query, predicates, and directives

wookayin opened the pull request:
Status

* Revisit after #27194 is merged
* Revisit after API changes on the predicates/directives #24738 (still other docs are useful)

---

Rewrite docs for treesitter query predicates and directives to be more
clear and comprehensive.

Revise lua type annotations: `TSPredicate` and `TSDirective`.

TSPredicate:
- handler function: fun(match, pattern, source, predicate): boolean
  - fix type for pattern: string -> integer
  - fix: buffer (integer) -> source
  - fix type for predicate: -> (string|integer)[]

TSDirective:
- handler function: fun(match, pattern, source, directive, metadata)
  - fix type for pattern: string -> integer
  - fix: buffer (integer) -> source (integer|string)
  - rename: predicate -> directive
  - fix type for directive: -> (string|integer)[]

TSQueryInfo:
  - rewrite docs for TSQueryInfo, explaining `info.captures` and
    `info.patterns`.
  - fix: key to table `info.patterns` (pattern) is integer, not string.

Improve docs for `vim.treesitter.query.get()` and
`vim.treesitter.query.set()`.

Add a test case for `vim.treesitter.query.add_directive()`.

Correct typing for `TSNode:_rawquery()`: `pattern` is integer, not
string.

Revise docs for `Query:iter_captures` and `Query:iter_matches`.

wookayin committed:
docs(treesitter): docs and typing for predicates and directives

Rewrite docs for treesitter query predicates and directives to be more
clear and comprehensive.

Revise lua type annotations: `TSPredicate` and `TSDirective`.

TSPredicate:
- handler function: fun(match, pattern, source, predicate): boolean
  - fix type for pattern: string -> integer
  - fix: buffer (integer) -> source
  - fix type for predicate: -> (string|integer)[]

TSDirective:
- handler function: fun(match, pattern, source, directive, metadata)
  - fix type for pattern: string -> integer
  - fix: buffer (integer) -> source (integer|string)
  - rename: predicate -> directive
  - fix type for directive: -> (string|integer)[]

Improve docs for `vim.treesitter.query.get()` and
`vim.treesitter.query.set()`.

Add a test case for `vim.treesitter.query.add_directive()`.

Correct typing for `TSNode:_rawquery()`: `pattern` is integer, not
string (used in predicates and directives).

Revise docs for `Query:iter_captures` and `Query:iter_matches`.

lucario387 commented on runtime/doc/treesitter.txt:
```diff
@@ -848,32 +848,55 @@ add_directive({name}, {handler}, {force})
-      • {name}     (string) Name of the directive, without leading #
-      • {handler}  function(match:table<string,|TSNode|>, pattern:string,
-                   bufnr:integer, predicate:string[], metadata:table)
+      • {name}     (string) Name of the directive, without leading `#`. Should
+                   end with `!`.
```
```suggestion
      • {name}     (string) Name of the directive, without leading `#`. Must
                   end with `!`.
```

See https://github.com/neovim/neovim/blob/master/runtime/lua/vim/treesitter/query.lua#L578

wookayin replied:
Good suggestion, I added that along with error checking.

lucario387 commented on runtime/lua/vim/treesitter/query.lua:
```diff
@@ -407,51 +419,53 @@ local predicate_handlers = {
+  -- |treesitter-directive-set!|
+  ['set!'] = function(_, _, _, directive, metadata)
+    if #directive >= 3 and type(directive[2]) == 'number' then
       -- (#set! @capture key value)
-      local capture_id, key, value = pred[2], pred[3], pred[4]
+      local capture_id, key, value = directive[2], directive[3], directive[4]
+      assert(type(capture_id) == 'number')
```
Doesn't the if condition already check for directive[2] to be number or not

wookayin replied:
you're right, it's unnecessary.

tomtomjhj commented on runtime/lua/vim/treesitter/query.lua:
```diff
@@ -3,14 +3,14 @@ local language = require('vim.treesitter.language')
 ---@class TSQueryInfo
 ---@field captures table
----@field patterns table<string,any[][]>
+---@field patterns table<integer,(integer|string)[][]> see |vim.treesitter.query.parse()|
```
`(integer|string)[][][]` (see `query_inspect`)

wookayin replied:
Not really; it might be correct if we say `(integer|string|nil)[][][]`, which is however confusing. The key (pattern_id) comes from the (1-based) index, in the order they appear in a query file, but they are not necessarily contiguous because some of them may not have predicates/directives. E.g.:

```lua
=vim.treesitter.query.get('c', 'locals').info
{
  captures = { "definition.function", "definition.macro", "scope", "definition.var", "definition.parameter", "definition.type", "definition.field", "definition", "reference" },
  patterns = {
    [15] = { { "set!", "reference.kind", "field" } },
    [16] = { { "set!", "reference.kind", "type" } }
  }
}
```
I will improve the documentation about the meaning of keys and values to be even clearer.

wookayin replied:
Updated

tomtomjhj commented on runtime/lua/vim/treesitter/query.lua:
TS API uses the term "index" instead of "id" for capture and pattern. For example, paramters of `ts_query_capture_quantifier_for_id` are named `capture_index` and `pattern_index`.

wookayin replied:
This is a good point. @lewis6991 Do we want to change and unify naming into `capture_index` and `pattern_index` from `capture_id`, `pattern_id`, etc.?

lewis6991 replied:
Yes, `id` is better.

wookayin commented:
Updates:

- More explanation about `TSQueryInfo.patterns`. Actually this should have been called as "predicates" (or directives), but it was there as a public API since #13260 (0.5.0) so we can't change them.
- Improve docs for `vim.treesitter.query.get()` and `vim.treesitter.query.set()`.

clason commented:
Treesitter is still declared experimental for this very reason: so we can make any changes if they're meaningful enough. If this is breaking actual use, we might want to add a migration path though.

wookayin commented:
For breaking changes (if they are necessary) I'll try to address in a separate PR.

lewis6991 commented on runtime/lua/vim/treesitter/query.lua:
```diff
@@ -370,8 +398,9 @@ local predicate_handlers = {
+  -- |treesitter-predicate-has-ancestor?|
   ['has-ancestor?'] = function(match, _, _, predicate)
-    local node = match[predicate[2]]
+    local node = match[predicate[2]] ---@type TSNode?
```
```suggestion
    local node = match[predicate[2] --[[@as integer]]]
```

wookayin replied:
`[` `[[..]]` `]` is a bit ugly so I chose to use another local variable.

wookayin commented:
Rebased to master and resolved conflicts. No changes in the docs since the last review.

wookayin committed:
docs(treesitter): add typing TSQuery, and document TSQueryInfo

- TSQuery: userdata object for parsed query.

- TSQueryInfo: document the fields (see TSQuery:inspect), and correct
  typing for `patterns`: should be a map from `integer` (pattern_id)
  to `(integer|string)[][]` (list of predicates/directives).

- Add typing for `vim._ts_parse_query()`.

wookayin commented on runtime/lua/vim/treesitter/query.lua:
```diff
@@ -1,16 +1,25 @@
----@field query userdata Parsed query
+---@field captures string[] List of (unique) capture names defined in query.
+---@field info TSQueryInfo Contains information used in queries, predicates, directives
+---@field query userdata Parsed query object (userdata)
```
Extracted part of this to: #27194

clason commented:
@wookayin What is the status here? I think our runtime is now fully typed (as far as luals is concerned), so only the doc improvements are left? It's probably better to add these in a fresh PR on top of `HEAD`.
